File: 2026/regional-andaluza/K/code/official.py
from collections import deque

def _bfs(capacity, source, sink, parent):
    queue = deque([source])
    visited = set([source])
    while queue:
        u = queue.popleft()
        for v in range(len(capacity)):
            if v not in visited and capacity[u][v] > 0:
                queue.append(v)
                visited.add(v)
                parent[v] = u
                if v == sink:
                    return True
    return False

def edmonds_karp(capacity, source, sink):
    max_flow = 0
    n = len(capacity)
    parent = [-1] * n

    while _bfs(capacity, source, sink, parent):
        path_flow = float('inf')
        v = sink
        while v != source:
            u = parent[v]
            path_flow = min(path_flow, capacity[u][v])
            v = u

        v = sink
        while v != source:
            u = parent[v]
            capacity[u][v] -= path_flow
            capacity[v][u] += path_flow
            v = u

        max_flow += path_flow

    return max_flow


def solve():
    T = int(input())
    for _ in range(T):
        N,M,K = map(int, input().split())
        V = 2 + M + K
        source = 0
        sink = V-1
        capacity = [[0 for _ in range(V)] for _ in range(V)]
        
        dicc = {}
        id = 0
        characters = input().split()
        for c in characters:
            id+=1
            dicc[c]=id 

        for i in range(1,M+1):
            capacity[source][i] = 1

        for i in range(1,M+1):
            c1,c2 = input().split()
            id1 = dicc[c1]
            id2 = dicc[c2]
            capacity[i][M+id1] = 1
            capacity[i][M+id2] =1

        class_limit = N//K
        for i in range(1,K+1):
            vol_node = M+i
            capacity[vol_node][sink] = class_limit

        flow = edmonds_karp(capacity, source, sink)
        print("YES" if flow==M else "NO")

if __name__ == "__main__":
    solve()

File: 2026/regional-andaluza/K/code/official.java
import java.util.*;

public class official {

    // El BFS y Edmonds-Karp se mantienen óptimos usando Lista de Adyacencia
    static boolean bfs(int[][] capacity, ArrayList<Integer>[] adj, int source, int sink, int[] parent) {
        int n = capacity.length;
        boolean[] visited = new boolean[n];
        Queue<Integer> queue = new ArrayDeque<>();

        queue.add(source);
        visited[source] = true;
        parent[source] = -1;

        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (int v : adj[u]) {
                if (!visited[v] && capacity[u][v] > 0) {
                    queue.add(v);
                    visited[v] = true;
                    parent[v] = u;
                    if (v == sink) return true;
                }
            }
        }
        return false;
    }

    static int edmondsKarp(int[][] capacity, ArrayList<Integer>[] adj, int source, int sink) {
        int n = capacity.length;
        int[] parent = new int[n];
        int maxFlow = 0;

        while (bfs(capacity, adj, source, sink, parent)) {
            int pathFlow = Integer.MAX_VALUE;
            int v = sink;
            while (v != source) {
                int u = parent[v];
                pathFlow = Math.min(pathFlow, capacity[u][v]);
                v = u;
            }
            v = sink;
            while (v != source) {
                int u = parent[v];
                capacity[u][v] -= pathFlow;
                capacity[v][u] += pathFlow;
                v = u;
            }
            maxFlow += pathFlow;
        }
        return maxFlow;
    }

    public static void main(String[] args) {
        // Usamos Scanner básico
        Scanner sc = new Scanner(System.in);
        
        if (!sc.hasNextInt()) return;
        int T = sc.nextInt();

        while (T-- > 0) {
            int N = sc.nextInt();
            int M = sc.nextInt();
            int K = sc.nextInt();

            int V = 2 + M + K;
            int source = 0;
            int sink = V - 1;

            int[][] capacity = new int[V][V];
            
            @SuppressWarnings("unchecked")
            ArrayList<Integer>[] adj = new ArrayList[V];
            for (int i = 0; i < V; i++) {
                adj[i] = new ArrayList<>();
            }

            HashMap<String, Integer> map = new HashMap<>();
            for (int i = 1; i <= K; i++) {
                map.put(sc.next(), i);
            }

            for (int i = 1; i <= M; i++) {
                capacity[source][i] = 1;
                adj[source].add(i);
                adj[i].add(source);
            }

            for (int i = 1; i <= M; i++) {
                String c1 = sc.next();
                String c2 = sc.next();

                int id1 = map.get(c1);
                int id2 = map.get(c2);

                capacity[i][M + id1] = 1;
                adj[i].add(M + id1);
                adj[M + id1].add(i);

                capacity[i][M + id2] = 1;
                adj[i].add(M + id2);
                adj[M + id2].add(i);
            }

            int classLimit = N / K;
            for (int i = 1; i <= K; i++) {
                int volNode = M + i;
                capacity[volNode][sink] = classLimit;
                adj[volNode].add(sink);
                adj[sink].add(volNode);
            }

            int flow = edmondsKarp(capacity, adj, source, sink);

            // Salida lenta caso por caso
            if (flow == M) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
        sc.close();
    }
}